### ISSUE TITLE
`read(filename, String)` is not type stable

### ISSUE THREAD

**Issue by xitology** [JuliaLang/julia#26419]
In 0.7-DEV, `readstring(filename)` is deprecated in favor of `read(filename, String)`. However, the latter is not type stable. `readstring(filename)` was type stable in 0.6.

```
julia> versioninfo()
Julia Version 0.7.0-DEV.4537
Commit 61bf9359e9 (2018-03-11 19:30 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-7600U CPU @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
Environment:

julia> readstring("/dev/null")
┌ Warning: `readstring(filename::AbstractString)` is deprecated, use `read(filename, String)` instead.
│   caller = top-level scope
└ @ Core :0
""

help?> read
  ...
  read(filename::AbstractString, String)
  Read the entire contents of a file as a string.
  ...

julia> Core.Compiler.return_type(read, (String, Type{String}))
Any
```

**Comment by JeffBezanson** [comment]
Probably same cause as #19096, perhaps #26050.

**Comment by vtjnash** [comment]
Not quite the same, since this is creating a closure too. If it was just slurping, that would be fixed now by #26826.
